File: Go/ExampleTests/production/domain.go
package production

// The entities and collections the Shopping Cart and Record Filter
// specifications describe.

// ---------------------------------------------------------------------------
// Records
// ---------------------------------------------------------------------------

type IDValue struct {
	ID    IDForm
	Value int
}

func NewIDValue(id IDForm, value int) IDValue { return IDValue{ID: id, Value: value} }

// RecordFilter holds IDValue records and sums the ones carrying a given ID.
type RecordFilter struct {
	entries []IDValue
}

func NewRecordFilter() *RecordFilter { return &RecordFilter{} }

func (r *RecordFilter) Add(entry IDValue) { r.entries = append(r.entries, entry) }

func (r *RecordFilter) SumByLabel(filterLabel IDForm) int {
	sum := 0
	for _, e := range r.entries {
		if e.ID == filterLabel {
			sum += e.Value
		}
	}
	return sum
}

// FahrenheitToCelsius truncates toward zero, so -40F comes out -40C.
func FahrenheitToCelsius(fahrenheit int) int { return (fahrenheit - 32) * 5 / 9 }

// ---------------------------------------------------------------------------
// Calculator
// ---------------------------------------------------------------------------

type Calculator struct{}

func NewCalculator() *Calculator          { return &Calculator{} }
func (c *Calculator) Add(a, b int) int    { return a + b }

// ---------------------------------------------------------------------------
// Shopping
// ---------------------------------------------------------------------------

type Address struct {
	Street SimpleText
	City   SimpleText
	State  SimpleText
	ZIP    SimpleText
}

type CatalogItem struct {
	Name  SimpleText
	Price Dollar
}

func NewCatalogItem(name SimpleText, price Dollar) CatalogItem {
	return CatalogItem{Name: name, Price: price}
}

const (
	CatalogMinimum = 0
	CatalogMaximum = 10000000
)

// Catalog is the items on offer, each with the price an order line is charged.
type Catalog struct {
	items []CatalogItem
}

func NewCatalog() *Catalog { return &Catalog{} }

func (c *Catalog) Add(item CatalogItem) { c.items = append(c.items, item) }
func (c *Catalog) Read() []CatalogItem  { return c.items }
func (c *Catalog) Size() int            { return len(c.items) }

func (c *Catalog) PriceFor(name SimpleText) (Dollar, bool) {
	for _, i := range c.items {
		if i.Name == name {
			return i.Price, true
		}
	}
	return Dollar{}, false
}

type OrderItem struct {
	Name      SimpleText
	Quantity  int
	Price     Dollar
	ItemTotal Dollar
}

func NewOrderItem(name SimpleText, quantity int, price, itemTotal Dollar) OrderItem {
	return OrderItem{Name: name, Quantity: quantity, Price: price, ItemTotal: itemTotal}
}

func CreateOrderItem(name SimpleText, quantity int, price Dollar) OrderItem {
	return NewOrderItem(name, quantity, price, price.Times(quantity))
}

// OrderItemFromCatalog looks the price up rather than being told it.
func OrderItemFromCatalog(catalog *Catalog, name SimpleText, quantity int) (OrderItem, bool) {
	price, ok := catalog.PriceFor(name)
	if !ok {
		return OrderItem{}, false
	}
	return CreateOrderItem(name, quantity, price), true
}

const (
	OrderItemCollectionMinimum = 0
	OrderItemCollectionMaximum = 100
)

type OrderItemCollection struct {
	items []OrderItem
}

func NewOrderItemCollection() *OrderItemCollection { return &OrderItemCollection{} }

func (o *OrderItemCollection) Add(item OrderItem) { o.items = append(o.items, item) }
func (o *OrderItemCollection) Read() []OrderItem  { return o.items }
func (o *OrderItemCollection) Size() int          { return len(o.items) }

func (o *OrderItemCollection) ComputeTotal() Dollar {
	total := Dollar{}
	for _, i := range o.items {
		total = total.Plus(i.ItemTotal)
	}
	return total
}

type ShoppingCart struct {
	Items           *OrderItemCollection
	Shipping        Dollar
	Discount        Dollar
	TotalPrice      Dollar
	ShippingAddress Address
	BillingAddress  Address
}

func NewShoppingCart(items *OrderItemCollection) *ShoppingCart {
	return &ShoppingCart{Items: items}
}

// --- the two business rules ------------------------------------------------

// ShippingCostFor is free once the order reaches $100, otherwise a flat $5.
func ShippingCostFor(totalPrice Dollar) Dollar {
	if totalPrice.Cents() >= 10000 {
		return DollarFromCents(0)
	}
	return DollarFromCents(500)
}

// DiscountFor is tiered: under $25 nothing, to $99.99 five percent, $100 up ten.
func DiscountFor(totalPrice Dollar) Percentage {
	switch c := totalPrice.Cents(); {
	case c >= 10000:
		return NewPercentage(10)
	case c >= 2500:
		return NewPercentage(5)
	default:
		return NewPercentage(0)
	}
}

// --- the Total Cart Price rule, over a bare item total ---------------------
//
// Stated as functions of the item total so the rule can be checked straight from
// its Examples table, which gives a TotalItems figure and no items. "Discount
// applied before shipping calculated", per that table's own note.

// DiscountAmountFor is the discount as money: the tiered percentage of the total.
func DiscountAmountFor(totalItems Dollar) Dollar {
	return totalItems.PercentOf(DiscountFor(totalItems))
}

// ShippingFor is judged on what the customer pays, so after the discount.
func ShippingFor(totalItems Dollar) Dollar {
	return ShippingCostFor(totalItems.Minus(DiscountAmountFor(totalItems)))
}

// TotalPriceFor is the item total, less the discount, plus shipping.
func TotalPriceFor(totalItems Dollar) Dollar {
	return totalItems.Minus(DiscountAmountFor(totalItems)).Plus(ShippingFor(totalItems))
}

// --- what this cart comes to -----------------------------------------------

// Subtotal is what the items come to before any discount or shipping.
func (s *ShoppingCart) Subtotal() Dollar { return s.Items.ComputeTotal() }

func (s *ShoppingCart) DiscountAmount() Dollar { return DiscountAmountFor(s.Subtotal()) }

func (s *ShoppingCart) ShippingCost() Dollar { return ShippingFor(s.Subtotal()) }

func (s *ShoppingCart) ComputeTotal() Dollar { return TotalPriceFor(s.Subtotal()) }

File: Swift/ExampleTests/production/Domain.swift
import Foundation

// The entities and collections the Shopping Cart and Record Filter
// specifications describe.

// ---------------------------------------------------------------------------
// Records
// ---------------------------------------------------------------------------

public struct IDValue: Equatable {
    public let id: IDForm
    public let value: Int

    public init(id: IDForm, value: Int) {
        self.id = id
        self.value = value
    }
}

/// Holds IDValue records and sums the ones carrying a given ID.
public class RecordFilter {
    private var entries: [IDValue] = []

    public init() {}

    public func add(_ entry: IDValue) { entries.append(entry) }

    public func sumByLabel(_ filterLabel: IDForm) -> Int {
        return entries.filter { $0.id == filterLabel }.reduce(0) { $0 + $1.value }
    }
}

/// Truncates toward zero, so -40F comes out -40C.
public func fahrenheitToCelsius(_ fahrenheit: Int) -> Int {
    return (fahrenheit - 32) * 5 / 9
}

// ---------------------------------------------------------------------------
// Calculator
// ---------------------------------------------------------------------------

public class Calculator {
    public init() {}
    public func add(_ a: Int, _ b: Int) -> Int { return a + b }
}

// ---------------------------------------------------------------------------
// Shopping
// ---------------------------------------------------------------------------

public struct Address: Equatable {
    public let street: SimpleText
    public let city: SimpleText
    public let state: SimpleText
    public let zip: SimpleText

    public init(street: SimpleText, city: SimpleText,
                state: SimpleText, zip: SimpleText) {
        self.street = street
        self.city = city
        self.state = state
        self.zip = zip
    }
}

public struct CatalogItem: Equatable {
    public let name: SimpleText
    public let price: Dollar

    public init(name: SimpleText, price: Dollar) {
        self.name = name
        self.price = price
    }
}

public let catalogMinimum = 0
public let catalogMaximum = 10000000

/// The items on offer, each with the price an order line is charged.
public class Catalog: Equatable {
    private var items: [CatalogItem] = []

    public init() {}

    public static func == (a: Catalog, b: Catalog) -> Bool { a.items == b.items }

    public func add(_ item: CatalogItem) { items.append(item) }
    public func read() -> [CatalogItem] { return items }
    public var size: Int { return items.count }

    public func priceFor(_ name: SimpleText) -> Dollar? {
        return items.first { $0.name == name }?.price
    }
}

public struct OrderItem: Equatable {
    public let name: SimpleText
    public let quantity: Int
    public let price: Dollar
    public let itemTotal: Dollar

    public init(name: SimpleText, quantity: Int, price: Dollar, itemTotal: Dollar) {
        self.name = name
        self.quantity = quantity
        self.price = price
        self.itemTotal = itemTotal
    }

    public static func create(name: SimpleText, quantity: Int, price: Dollar) -> OrderItem {
        return OrderItem(name: name, quantity: quantity,
                         price: price, itemTotal: price.times(quantity))
    }

    /// Looks the price up rather than being told it.
    public static func fromCatalog(_ catalog: Catalog, name: SimpleText,
                                   quantity: Int) -> OrderItem? {
        guard let price = catalog.priceFor(name) else { return nil }
        return create(name: name, quantity: quantity, price: price)
    }
}

public let orderItemCollectionMinimum = 0
public let orderItemCollectionMaximum = 100

public class OrderItemCollection: Equatable {
    private var items: [OrderItem] = []

    public init() {}

    public static func == (a: OrderItemCollection, b: OrderItemCollection) -> Bool {
        a.items == b.items
    }

    public func add(_ item: OrderItem) { items.append(item) }
    public func read() -> [OrderItem] { return items }
    public var size: Int { return items.count }

    public func computeTotal() -> Dollar {
        return items.reduce(Dollar(cents: 0)) { $0.plus($1.itemTotal) }
    }
}

public class ShoppingCart {
    public let items: OrderItemCollection
    public let shippingAddress: Address?
    public let billingAddress: Address?

    public init(items: OrderItemCollection,
                shippingAddress: Address? = nil,
                billingAddress: Address? = nil) {
        self.items = items
        self.shippingAddress = shippingAddress
        self.billingAddress = billingAddress
    }

    // --- the two business rules -------------------------------------------

    /// Free once the order reaches $100, otherwise a flat $5.
    public static func shippingCostFor(_ totalPrice: Dollar) -> Dollar {
        return totalPrice.cents >= 10_000 ? Dollar(cents: 0) : Dollar(cents: 500)
    }

    /// Tiered: under $25 nothing, to $99.99 five percent, $100 up ten.
    public static func discountFor(_ totalPrice: Dollar) -> Percentage {
        if totalPrice.cents >= 10_000 { return Percentage(10) }
        if totalPrice.cents >= 2_500 { return Percentage(5) }
        return Percentage(0)
    }

    // --- the Total Cart Price rule, over a bare item total -----------------
    //
    // Stated as functions of the item total so the rule can be checked straight
    // from its Examples table, which gives a TotalItems figure and no items.
    // "Discount applied before shipping calculated", per that table's own note.

    /// The discount as money: the tiered percentage of the item total.
    public static func discountAmountFor(_ totalItems: Dollar) -> Dollar {
        return totalItems.percentOf(discountFor(totalItems))
    }

    /// Shipping is judged on what the customer pays, so after the discount.
    public static func shippingFor(_ totalItems: Dollar) -> Dollar {
        return shippingCostFor(totalItems.minus(discountAmountFor(totalItems)))
    }

    /// Item total, less the discount, plus shipping.
    public static func totalPriceFor(_ totalItems: Dollar) -> Dollar {
        return totalItems.minus(discountAmountFor(totalItems))
                         .plus(shippingFor(totalItems))
    }

    // --- what this cart comes to ------------------------------------------

    /// What the items come to before any discount or shipping.
    public func subtotal() -> Dollar { return items.computeTotal() }

    public func discountAmount() -> Dollar {
        return ShoppingCart.discountAmountFor(subtotal())
    }

    public func shippingCost() -> Dollar {
        return ShoppingCart.shippingFor(subtotal())
    }

    public func computeTotal() -> Dollar {
        return ShoppingCart.totalPriceFor(subtotal())
    }
}
